### PR TITLE
FastText Language Fallback Classifier & Printer Fix

### PR DESCRIPTION
**Pull request by koaning**
If you configure something like this; 

```yaml
language: en

pipeline:
- name: WhitespaceTokenizer
- name: LexicalSyntacticFeaturizer
- name: CountVectorsFeaturizer
  analyzer: char_wb
  min_ngram: 1
  max_ngram: 4
- name: DIETClassifier
  epochs: 1
- name: rasa_nlu_examples.classifiers.FasttextLanguageFallbackClassifier
  expected_language: en
  intent_triggered: non_english
  cache_dir: downloaded
  file: 'lid.176.ftz'
  min_chars: 5
  min_tokens: 2
  threshold: 0.3
```

Then you should be able to trigger a rule with it whenever a language is uttered that you don't support. The rule might look something like; 

```yaml
rules:

- rule: Ask the user to switch to speaking in English.
  steps:
  - intent: non_english
  - action: utter_non_english
```

### CODE REVIEW

**Comment by koaning** [comment]
Before merging it would be good to add a setting such that we prevent certain intents to be overwritten.

**Comment by koaning** [comment]
@tabergma The PR got a bit "big" unfortunately. The `printer` tool was broken because of a Rasa 2.0 compatibility bug that I overlooked earlier. The main thing that I need a review on is the `FasttextLanguageFallbackClassifier` and the associated tests. I mainly want to make sure that I am overriding the intent in the right way.

**Comment by koaning** [comment]
This PR also fixes https://github.com/RasaHQ/rasa-nlu-examples/issues/27

File: docs/index.md [hunk 1]
```diff
@@ -64,6 +67,37 @@ It's typically only about 5 lines of code. That means that you could train your
...
+
+**`rasa_nlu_examples.fallback.FasttextLanguageFallbackClassifier` [docs](docs/fallback/fasttextlanguagefallback.md)**
+
+This fallback classifier is based on [fasttext](https://fasttext.cc/docs/en/language-identification.html). It
+can detect when a user is speaking in an unintended language such that you can create a rule to respond
+appropriately.
+
+## **Meta**
+
+The components listed here won't effect the NLU pipeline but they might instead might cause extra logs
```
Comment (tabergma) [inline docs/index.md:90]:
```suggestion
The components listed here won't effect the NLU pipeline but they might instead cause extra logs
```

File: docs/docs/meta/printer.md [hunk 2]
```diff
@@ -27,3 +27,46 @@ pipeline:
 - name: DIETClassifier
   epochs: 100
 ```
+
+When you now interactive with a trained NLU model from the terminal you should see pretty output appear in
+the terminal. It should look look the output below.
```
Comment (tabergma) [inline docs/docs/meta/printer.md:32]: This sentence is weird. Can you please rephrase?

File: docs/docs/fallback/fasttextlanguagefallback.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+```python
+import fasttext
+model = fasttext.load_model("lid.176.ftz")
+
+txt = "i am speaking english"
+proba_dict = {k: v for k, v in zip(*model.predict(txt, k=10))}
+proba_dict
+```
+
+## Configurable Variables
```
Comment (tabergma) [inline docs/docs/fallback/fasttextlanguagefallback.md:34]: You might want to add the default values.

File: rasa_nlu_examples/fallback/fasttext_language_fallback.py [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+            )
+        path = os.path.join(component_config["cache_dir"], component_config["file"])
+        if not os.path.exists(component_config["cache_dir"]):
+            raise FileNotFoundError(
+                f"It seems that the cache dir {component_config['cache_dir']} does not exists. Please check config."
+            )
+        if not os.path.exists(path):
+            raise FileNotFoundError(
+                f"It seems that file {path} does not exists. Please check config."
+            )
```
Comment (tabergma) [inline rasa_nlu_examples/fallback/fasttext_language_fallback.py:250]: I would put all of this into an extra method called `_validate_config`.

File: rasa_nlu_examples/fallback/fasttext_language_fallback.py [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+        self.expected_language = component_config["expected_language"]
+        self.min_chars = component_config["min_chars"]
+        self.min_tokens = component_config["min_tokens"]
+        self.threshold = component_config["threshold"]
+        self.intent_triggered = component_config["intent_triggered"]
+        self.protected_intents = (
+            component_config["protected_intents"]
+            if "protected_intents" in component_config
+            else self.defaults["protected_intents"]
+        )
```
Comment (tabergma) [inline rasa_nlu_examples/fallback/fasttext_language_fallback.py:262]: `protected_intents` should always be in `component_config` as the default values are added to the `component_config` during init of the component.

Comment (tabergma) [reply rasa_nlu_examples/fallback/fasttext_language_fallback.py:262]:
```suggestion
        self.protected_intents = component_config["protected_intents"]
```

File: rasa_nlu_examples/fallback/fasttext_language_fallback.py [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+            return
+
+        if self._not_enough_text(message):
+            logger.debug(
+                "There's not enough text to do proper language identification."
+            )
+            return
+
+        proba_dict = {
+            k: v for k, v in zip(*self.model.predict(message.get(TEXT), k=20))
```
Comment (tabergma) [inline rasa_nlu_examples/fallback/fasttext_language_fallback.py:281]: I guess `k` refers to the language code and `v` to the confidence value? If so, I would rename the variables accordingly.

File: rasa_nlu_examples/meta/printer.py [hunk 7]
```diff
@@ -13,39 +15,41 @@
...
-      alias: before count vectors
-    - name: CountVectorsFeaturizer
-      analyzer: char_wb
-      min_ngram: 1
-      max_ngram: 4
-    - name: rasa_nlu_examples.meta.Printer
-      alias: after count vectors
-    - name: DIETClassifier
-      epochs: 1
+def print_message(msg: Message):
```
Comment (tabergma) [inline rasa_nlu_examples/meta/printer.py:26]:
```suggestion
def print_message(message: Message) -> None:
```

File: tests/test_fallback/test_ft_lang_fallback.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    message = Message({TEXT: txt, INTENT: {"name": "assigned_before"}})
+    tokenizer.process(message=message)
+    ft_lang.process(message=message)
+    assert message.get(INTENT)["name"] == "assigned_before"
+
+
+@pytest.mark.parametrize(
+    "txt,lang", [("ik spreek een taal", "nl"), ("je parle une langue", "fr")]
+)
+def test_no_change_when_protected(txt, lang):
```
Comment (tabergma) [inline tests/test_fallback/test_ft_lang_fallback.py:53]: What about a test that nothing changes when text is too short?

File: rasa_nlu_examples/meta/printer.py [hunk 9]
```diff
@@ -13,39 +15,41 @@
     from rasa.nlu.model import Metadata
 
 
-class Printer(Component):
-    """
-    A component that prints the message. Useful for debugging while running `rasa shell`.
-
-    **Configurable Variables:**
+def dense_msg(arr: np.ndarray):
```
Comment (tabergma) [inline rasa_nlu_examples/meta/printer.py:18]:
```suggestion
def dense_message(dense_array: np.ndarray) -> Dict[Text, Any]:
```

File: rasa_nlu_examples/meta/printer.py [hunk 10]
```diff
@@ -13,39 +15,41 @@
...
+    return {"shape": arr.shape, "dtype": arr.dtype}
 
-    - alias: gives an extra name to the component and adds an extra message that is printed
 
-    Usage:
+def sparse_msg(arr: scipy.sparse.spmatrix):
```
Comment (tabergma) [inline rasa_nlu_examples/meta/printer.py:22]:
```suggestion
def sparse_message(sparse_array: scipy.sparse.spmatrix) -> Dict[Text, Any]:
```

**Review by tabergma — APPROVED** [review]
Overall it looks good 🚀  The intent is set correctly. Left a few comments.

**Pull request merged**